### PR TITLE
Adding support for loop-telecom devices

### PR DESCRIPTION
**Pull request by Blinq-SanderBlom**
Adding support for telecom products from Loop-Telecom. 

Hardware setup:
- AM3440-CCPB-A with 3 ethernet ports (SNMP, ETH 1 & ETH 2
- 2x 150W-dc psu 
- 10x fxs cards
- 1x CLKa card 

What's working
- Initial os detection 
- Reading current power draw
- Detecting Ethernet ports (speed, mode (half/full duplex), packets in, out and errors)
- Read chassi temprature
- Status for all line cards insert.
- Status for PSUs
- Console, telnet & ssh access status 
- Misc voltage & current draw

What's not working
- ~~Collecting packet count and graphing port traffic (think I'm collecting the data correctly. Do i have to I have to define that I want to create graphs for each port?)~~
- ~~States for PSU and line cards stops working after the poller run for the first time. If I delete and add the device again the states are working again. Looks the the `{{$index}} ` for the linecars num_oid reports the slot name insted of slot number. Not sure if i can fix this in the yaml file or if I have to do the integration with php instead.~~
- ~~Hardware name includes the name of the mib ( L-AM3440-A-Private::loop-AM3440-CCPB-A. should be just "loop-AM3440-CCPB-A")~~
- ~~Problems saving test data~~

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by Jellyfrog** [comment]
Mark it as a draft pr instead

File: includes/discovery/sensors/temperature/loop-telecom.inc.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
```
Comment (PipoCanaja) [inline includes/discovery/sensors/temperature/loop-telecom.inc.php:1]: Hi,
You can probably do it in YAML and provide snmp_hexstring as user-func.

**Comment by PipoCanaja** [comment]
Hi
Please rename the mibfile to `L-AM3440-A-Private` as this is the official MIB definition. 
Thanx

File: includes/discovery/sensors/temperature/loop-telecom.inc.php (hunk 1 above)
Comment (Blinq-SanderBlom) [reply includes/discovery/sensors/temperature/loop-telecom.inc.php:1]: Could you provide me with an example on how to do this? 
I could not find this anywhere in the documentation. 

Thank you!

**Comment by Blinq-SanderBlom** [comment]
@Jellyfrog  any idea why the `save-test-data.php` fails ?

File: includes/definitions/discovery/loop-telecom.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+mib: L-AM3440-A-Private
+modules:
+  os:
+    hardware: SNMPv2-MIB::sysObjectID.0
```
Comment (murrant) [inline includes/definitions/discovery/loop-telecom.yaml:4]: Delete this line and set `hardware_mib: L-AM3440-A-Private` instead.

Comment (murrant) [reply includes/definitions/discovery/loop-telecom.yaml:4]: https://docs.librenms.org/Developing/os/Initial-Detection/#os-discovery

File: includes/polling/ports/os/loop-telecom.inc.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            $port_stats[$curIfIndex]['ifSpeed'] = 10000000;
+            $port_stats[$curIfIndex]['ifDuplex'] = 'fullDuplex';
+            break;
+        case 7: //10mbps half duplex
+            $port_stats[$curIfIndex]['ifSpeed'] = 10000000;
+            $port_stats[$curIfIndex]['ifDuplex'] = 'halfDuplex';
+            break;
+    }
+    //Set port mode
+    switch ($port['ethernetStatusMode']) {
```
Comment (murrant) [inline includes/polling/ports/os/loop-telecom.inc.php:51]: These are not valid ifType values.  See IANAifType-MIB

Comment (murrant) [reply includes/polling/ports/os/loop-telecom.inc.php]: That would probably be in inventory (usually supplied by ENTITY-MIB)

**Comment by murrant** [comment]
There is a variable for that that converts strings into numeric oids called $index_string

**Comment by murrant** [comment]
You might want to try `-OQUsb`

https://linux.die.net/man/1/snmpcmd

**Comment by Blinq-SanderBlom** [comment]
Trying to get the temp reading working with yaml but I'm still having issues. The closest thing i get is converting the hex to a string. Current output:  `60.XC`. Is there any way to keep just the integers?  @murrant 

My yaml atm 

```
    temperature:
      data:
        - oid: c1Temperature
          num_oid: '.1.3.6.1.4.1.823.34441.1.10.11.0'
          snmp_flags: '-OQUsb'
          descr: 'Temp sensor n'
          user_func: snmp_hexstring 

```

Logs: 

```
SNMP[[0;36m'/usr/bin/snmpget' '-v1' '-c' 'COMMUNITY' '-OUQntea' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/loop-telecom' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.823.34441.1.10.11.0'[0m]
.*.4.1.823.344*.0 = " 60.XC"  

```

**Comment by murrant** [comment]
You might try updating Number::cast() like this:
```php
        if (! is_numeric($number)) {
            // match pre-PHP8 behavior
            if (! preg_match('/^\s*(-?\d+(\.\d+)?)/', $number ?? '', $matches)) {
                return 0;
            }
            $number = $matches[1];
        }
```
to ignore spaces before the number and see if it breaks any other ones when the tests run.

**Comment by Blinq-SanderBlom** [comment]
@murrant. I got the temperature reading to work using yaml but I was not successful with the states because of the issues mentioned in previous comments. Any idea on why i can't save the test data and why the mysql & mariadb tests fails?

**Comment by murrant** [comment]
FYI, I merged #14699 which will handle whitespace in front of the number.

**Comment by Blinq-SanderBlom** [comment]
@murrant @PipoCanaja any thing else needed to merge the pull request?

**Comment by Jellyfrog** [comment]
Missing json test data

**Comment by Blinq-SanderBlom** [comment]
> Missing json test data

Json test data has now been added :)

**Comment by Jellyfrog** [comment]
There is no other name for the operating system? (librenms works with OSes rather than vendors)

**Review by Jellyfrog — APPROVED** [review]

**Comment by Blinq-SanderBlom** [comment]
> There is no other name for the operating system? (librenms works with OSes rather than vendors)

Unfortunately not. I have not found any mentioning of the OS name anywhere in the device or other devices I have used from the same vendor the last 5 years. Therefor I assume they use the same OS on all devices.

**Comment by Jellyfrog** [comment]
Good job here!

**Pull request merged**